src/coreclr/jit/importer.cpp

```diff
@@ -9497,15 +9495,67 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                                                                                   genTypeSize(lclTyp)))
                                 {
-                                    GenTree* cnsValue = impImportStaticReadOnlyField(buffer, bufferSize, lclTyp);
+                                    GenTree* cnsValue = impImportStaticReadOnlyField(buffer, lclTyp);
                                     if (cnsValue != nullptr)
                                     {
                                         op1 = cnsValue;
                                         goto FIELD_DONE;
                                     }
                                 }
                             }
+                            else if (lclTyp == TYP_STRUCT)
+                            {
+                                CORINFO_CLASS_HANDLE fieldClsHnd;
+                                info.compCompHnd->getFieldType(resolvedToken.hField, &fieldClsHnd,
+                                                               resolvedToken.hClass);
+
+                                if (info.compCompHnd->getClassNumInstanceFields(fieldClsHnd) != 1)
+                                {
+                                    // Only single-field structs are supported here to avoid potential regressions where
+                                    // Metadata-driven struct promotion leads to regressions.
+                                    goto NONFOLDABLE_FIELD;
+                                }
+
+                                const int bufferSize         = TARGET_POINTER_SIZE;
+                                uint8_t   buffer[bufferSize] = {0};
+                                unsigned  totalSize          = info.compCompHnd->getClassSize(fieldClsHnd);
+
+                                if ((totalSize == 0) || (totalSize > bufferSize) ||
+                                    !info.compCompHnd->getReadonlyStaticFieldValue(resolvedToken.hField, buffer,
+                                                                                   totalSize))
+                                {
+                                    goto NONFOLDABLE_FIELD;
+                                }
+
+                                CORINFO_FIELD_HANDLE innerField = info.compCompHnd->getFieldInClass(fieldClsHnd, 0);
+                                CORINFO_CLASS_HANDLE innerFieldClsHnd;
+                                var_types            filedVarType = JITtype2varType(
+                                    info.compCompHnd->getFieldType(innerField, &innerFieldClsHnd, fieldClsHnd));
+
+                                // Technically, we can support frozen gc refs here and maybe floating point in future
+                                if (!varTypeIsIntegral(filedVarType))
+                                {
+                                    goto NONFOLDABLE_FIELD;
+                                }
+
+                                unsigned fldOffset     = info.compCompHnd->getFieldOffset(innerField);
```

Comment:
Seems a bit odd to limit this to single field structs and then not check that the field is the entire struct (offset 0, same size).

Reply:
@AndyAyersMS thanks! Added a check for size + offset in https://github.com/dotnet/runtime/pull/77102/commits/0bf8917ab73b600d9429e660411203694ba05443 commit